can/io/blf.py

```diff
@@ -0,0 +1,263 @@
+"""
+Implements support for BLF (Binary Logging Format) which is a proprietary
+CAN log format from Vector Informatik GmbH.
+
+No offical specification is available. The implementation is based on the works
+by Toby Lorenz (https://bitbucket.org/tobylorenz/vector_blf).
+The file starts with a header. The rest is one or more "log containers"
+which consists of a header and some zlib compressed data, usually up to 128 kB
+of uncompressed data each. This data contains the actual CAN messages and other
+objects types.
+"""
+import logging
+import struct
+import zlib
+import datetime
+import time
+
+from can.message import Message
+from can.CAN import Listener
+
+
+logger = logging.getLogger(__name__)
+
+# 0 = unknown, 2 = CANoe
+APPLICATION_ID = 5
+
+# Header must be 144 bytes in total
+# signature ("LOGG"), header size,
+# application ID, application major, application minor, application build,
+# bin log major, bin log minor, bin log build, bin log patch,
+# file size, uncompressed size, count of objects, count of objects read,
+# time start (SYSTEMTIME), time stop (SYSTEMTIME)
+FILE_HEADER_STRUCT = struct.Struct("<4sLBBBBBBBBQQLL8H8H72x")
+
+# signature ("LOBJ"), header size, header version (1), object size, object type,
+# flags, object version, size uncompressed or timestamp
+OBJ_HEADER_STRUCT = struct.Struct("<4sHHLLL2xHQ")
+
+# channel, flags, dlc, arbitration id, data
+CAN_MSG_STRUCT = struct.Struct("<HBBL8s")
+
+# channel, length
+CAN_ERROR_STRUCT = struct.Struct("<HH4x")
+
+# commented event type, foreground color, background color, relocatable,
+# group name length, marker name length, description length
+GLOBAL_MARKER_STRUCT = struct.Struct("<LLL3xBLLL12x")
+
+
+CAN_MESSAGE = 1
+CAN_ERROR = 2
+LOG_CONTAINER = 10
+GLOBAL_MARKER = 96
+
+CAN_MSG_EXT = 0x80000000
+REMOTE_FLAG = 0x80
+
+
+def timestamp_to_systemtime(timestamp):
+    t = datetime.datetime.fromtimestamp(timestamp)
+    return (t.year, t.month, t.isoweekday() % 7, t.day,
+            t.hour, t.minute, t.second, int(round(t.microsecond / 1000.0)))
+
+
+def systemtime_to_timestamp(systemtime):
+    try:
+        t = datetime.datetime(
+            systemtime[0], systemtime[1], systemtime[3],
+            systemtime[4], systemtime[5], systemtime[6], systemtime[7] * 1000)
+        return time.mktime(t.timetuple()) + systemtime[7] / 1000.0
+    except ValueError:
+        return 0
+
+
+class BLFReader(object):
+    """
+    Iterator of CAN messages from a Binary Logging File.
```

Comment:
Probably worth mentioning that any non CAN message objects are dropped when encountered.